specs/withdrawals.md

```diff
@@ -0,0 +1,176 @@
+
+Withdrawals are transactions initiated on L2, and executed on L1. They may be used to transfer data, ether or both from
+L1 to L2.
+
+**Vocabulary note**:
+
+- _withdrawing transaction_ refers specifically to an L2 transaction
+- _withdrawn transaction_ refers specifically to an L1 transaction
+- _withdrawal_ can refer to the transaction at various stages
+
+## Constants
+
+| Name                  | Value     | Unit    |
+| --------------------- | --------- | ------- |
+| `FINALIZATION_WINDOW` | `604_800` | seconds |
+
+This `FINALIZATION_WINDOW` value is equivalent to 7 days.
+
+## Data Structures and Type Aliases
+
+1. A `WithdrawalMessage` is encoded in a struct as follows:
+
+    ```js
+    struct WithdrawalMessage {
+        uint256 nonce;
+        address sender;
+        uint256 value;
+        bytes message;
+    }
+    ```
+
+1. The `L2OutputTimestamp` is an alias for `uint256`, and MUST be a multiple of the `SUBMISSION_INTERVAL` described
+  in the [L2 Output](./proposals.md#constants) document.
+
+1. The `WithdrawalsRoot` is an alias for a `bytes32` value, corresponding to the [MPT][g-mpt]
+  storage root of the Withdrawals predeploy contract at `0x4200000000000000000000000000000000000015` (described below).
+
+1. The `WithdrawalsRootInclusionProof` proof contains the data necessary to prove that the provided `WithdrawalsRoot` is
+  included in the [SSZ merkleization](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization)
+  of the `L2Output` as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+1. A `WithdrawalMessageInclusionProof` is an MPT proof encoded in a struct as follows:
+
+   ```js
+   struct WithdrawalMessageInclusionProof {
+       WithdrawalMessage message;
+       bytes32 l2withdrawalsRoot;
+       bytes memory _key;      // storage key of the withdrawal message commitment
+       bytes memory _value;    // Always bytes32(1) (boolean validity status of the message)
+       bytes memory _proof;    // MPT inclusion proof for the key/value
+   }
+   ```
+1. This document also refers to the `L2Output` type as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+## Withdrawal verification
+
+In order to verify a `WithdrawalMessage`, the following inputs must be provided:
+
+| Type                              | Name               |
+| --------------------------------- | ------------------ |
+| `WithdrawalMessage`               | `message`          |
+| `L2OutputTimestamp`               | `timestamp`        |
+| `WithdrawalsRootInclusionProof`   | `storageRootProof` |
+| `WithdrawalMessageInclusionProof` | `messageProof`     |
+
+These inputs must satisfy the following conditions:
+
+1. The `timestamp` is at least `FINALIZATION_WINDOW` seconds old.
+1. `OutputOracle.l2Outputs(timestamp)` returns a non-zero value `l2Output`.
```

Comment:
`OutputOracle.l2Outputs(timestamp)` probably needs to be updated to reflect the introduction of a timestamp which is recorded at time of submission